vault/identity_store_oidc_provider.go

```diff
@@ -1771,6 +1815,25 @@ func (i *IdentityStore) pathOIDCToken(ctx context.Context, req *logical.Request,
 		return tokenResponse(nil, ErrTokenInvalidRequest, "identity entity not authorized by client assignment")
 	}
 
+	// Validate the code verifier if a code challenge and code challenge
+	// method are associated with the authorization code. See details at
+	// https://datatracker.ietf.org/doc/html/rfc7636#section-4.6
+	if authCodeEntry.codeChallenge != "" && authCodeEntry.codeChallengeMethod != "" {
```

Comment:
It seems like we are still requiring the Client Secret even when using PKCE. I can't find it explicitly stated in the spec, but it doesn't seem like the client secret should be required when using PKCE.

Reply:
Good observation! I was just looking into this :) Doing some research to see what we should do.

Reply:
Okay, we shouldn't require the client secret if PKCE is being used, since PKCE is _effectively_ the auth method for public clients that can't protect their client secret. 

As an approach, I'm thinking we should define a configuration field for clients, `token_endpoint_auth_method`. The name aligns with client metadata defined in [dynamic client registration](https://openid.net/specs/openid-connect-registration-1_0.html). The field would support the following values for [client authentication](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication):
- `client_secret_basic`
  - This is the default that we already support in the code
- `none`
  - We would enforce that PKCE be used if `none` is selected
  - We likely don't need to return/issue client secrets for clients using `none`

We could then add support for different token auth methods in the future. Happy to consider any thoughts around this (cc: @calvn).

Reply:
Yep, that makes sense to me!

Reply:
> We likely don't need to return/issue client secrets for clients using none

Can a client change its token auth method after creation, say from PKCE back to basic auth?

Reply:
I'm going to resolve this discussion as some of the information is stale from conversations outside of this PR. We will allow clients to explicitly declare as public/confidential. The client's won't be able to switch between the two (i.e., set at creation time and can't be updated). We'll require PKCE for public clients that use the authorization code flow. We'll allow confidential clients to use PKCE, but it won't be required.